elasticgraph-graphql/lib/elastic_graph/graphql/filtering/filter_interpreter.rb

```diff
@@ -313,6 +311,9 @@ def process_list_count_expression(bool_node, expression, field_path)
           # While we index an explicit count of 0, the count field will be missing from documents indexed before
           # the list field was defined on the ElasticGraph schema. To properly match those documents, we need to
           # convert this into an OR (using `any_of`) to also match documents that lack the field entirely.
+          expression = expression.compact
+          return if expression.empty?
```

Comment:
This worries me a little--it passes along a compacted `expression` to `process_sub_field_expression` which in turn turns around and calls `process_filter_hash` which is where we're trying to remove the compacting from!  I haven't been able to think of a concrete example of a bug that it would cause but I'm more comfortable with a `compact` if it is obviously isolated to within a pure function (that just uses it to return a value) rather than changing the expression passed along as we recurse.

I played around locally and came up with an alternative--WDYT?

```diff
diff --git a/elasticgraph-graphql/lib/elastic_graph/graphql/filtering/filter_interpreter.rb b/elasticgraph-graphql/lib/elastic_graph/graphql/filtering/filter_interpreter.rb
index 1de266e..5838802 100644
--- a/elasticgraph-graphql/lib/elastic_graph/graphql/filtering/filter_interpreter.rb
+++ b/elasticgraph-graphql/lib/elastic_graph/graphql/filtering/filter_interpreter.rb
@@ -311,10 +311,7 @@ module ElasticGraph
           # While we index an explicit count of 0, the count field will be missing from documents indexed before
           # the list field was defined on the ElasticGraph schema. To properly match those documents, we need to
           # convert this into an OR (using `any_of`) to also match documents that lack the field entirely.
-          expression = expression.compact
-          return if expression.empty?
-
-          unless excludes_zero?(expression)
+          if filters_to_range_including_zero?(expression)
             expression = {schema_names.any_of => [
               expression,
               {schema_names.equal_to_any_of => [nil]}
@@ -338,20 +335,28 @@ module ElasticGraph
           {bool: bool_node}
         end
 
-        # Determines if the given filter expression excludes the value `0`.
-        def excludes_zero?(expression)
-          expression.any? do |operator, operand|
-            case operator
-            when schema_names.equal_to_any_of then !operand.include?(0)
-            when schema_names.lt then operand <= 0
-            when schema_names.lte then operand < 0
-            when schema_names.gt then operand >= 0
-            when schema_names.gte then operand > 0
-            else
-              # :nocov: -- all operators are covered above. But simplecov complains about an implicit `else` branch being uncovered, so here we've defined it to wrap it with `:nocov:`.
-              false
-              # :nocov:
-            end
+        # Determines if the given expression filters to a range that includes `0`.
+        # If it does not do any filtering (e.g. an empty expression) it will return `false`.
+        def filters_to_range_including_zero?(expression)
+          expression = expression.compact
+
+          expression.size > 0 && expression.none? do |operator, operand|
+            operator_excludes_zero?(operator, operand)
+          end
+        end
+
+        # Determines if the given operator and operand exclude 0 as a matched value.
+        def operator_excludes_zero?(operator, operand)
+          case operator
+          when schema_names.equal_to_any_of then !operand.include?(0)
+          when schema_names.lt then operand <= 0
+          when schema_names.lte then operand < 0
+          when schema_names.gt then operand >= 0
+          when schema_names.gte then operand > 0
+          else
+            # :nocov: -- all operators are covered above. But simplecov complains about an implicit `else` branch being uncovered, so here we've defined it to wrap it with `:nocov:`.
+            false
+            # :nocov:
           end
         end
 ```

This feels safer (and easier to reason about) to me. The `compact` is just used within `filters_to_range_including_zero?` to determine if a range filter is being applied at all, and the `compact` is clearly isolated within that method and doesn't "leak" into the recursion to impact something later.